File: src/components/ResizableCanvas/snapLogic.ts
import { BoundingBox, Guideline, Polygon } from './types';

interface SnapResult {
  translation: { x: number; y: number };
  guidelines: Guideline[];
}

// Helper to check snap
const checkSnap = (
    value: number,
    target: number,
    type: 'vertical' | 'horizontal',
    start: number,
    end: number,
    threshold: number,
    currentMinDist: number
  ) => {
    const dist = Math.abs(value - target);
    if (dist < currentMinDist && dist < threshold) {
      return { 
        dist, 
        snapDelta: target - value, 
        match: { type, pos: target, start, end } 
      };
    }
    return null;
  };
  
  export const calculateSnap = (
    proposedBox: BoundingBox,
    polygons: Polygon[],
    excludedIds: number[],
    canvasWidth: number,
    canvasHeight: number,
    threshold: number
  ): SnapResult => {
    
    // Filter and calculate boxes for other polygons
    const otherBoxes = polygons
      .map((p, i) => ({ p, i }))
      .filter(({ i }) => !excludedIds.includes(i))
      .map(({ p }) => {
        // Simplified bounding box calculation for testability/reuse if points are strings
        const points = p.points.split(' ').flatMap(pair => {
            const [x, y] = pair.split(',').map(Number);
            return [{x, y}];
        });
        const xs = points.map(pt => pt.x);
        const ys = points.map(pt => pt.y);
        return {
            x: Math.min(...xs),
            y: Math.min(...ys),
            width: Math.max(...xs) - Math.min(...xs),
            height: Math.max(...ys) - Math.min(...ys),
        } as BoundingBox;
      });
  
    // Add canvas edges
    otherBoxes.push(
      { x: 0, y: 0, width: canvasWidth, height: canvasHeight }
    );
  
    const newGuidelines: Guideline[] = [];
    let snapDeltaX = 0;
    let snapDeltaY = 0;
    let minDistX = threshold;
    let minDistY = threshold;
  
    const edgesX = [
      { val: proposedBox.x, type: 'start' }, 
      { val: proposedBox.x + proposedBox.width / 2, type: 'center' }, 
      { val: proposedBox.x + proposedBox.width, type: 'end' }
    ];
    const edgesY = [
      { val: proposedBox.y, type: 'start' }, 
      { val: proposedBox.y + proposedBox.height / 2, type: 'center' }, 
      { val: proposedBox.y + proposedBox.height, type: 'end' }
    ];
  
    type SnapMatch = { target: number, start: number, end: number };
    let bestXMatch: SnapMatch | null = null;
    let bestYMatch: SnapMatch | null = null;
  
    for (const target of otherBoxes) {
        const tX = [target.x, target.x + target.width / 2, target.x + target.width];
        const tY = [target.y, target.y + target.height / 2, target.y + target.height];
  
        // Vertical Guides (Horizontal movement)
        for (const edge of edgesX) {
            for (const targetVal of tX) {
                const result = checkSnap(edge.val, targetVal, 'vertical', 
                    Math.min(proposedBox.y, target.y), 
                    Math.max(proposedBox.y + proposedBox.height, target.y + target.height),
                    threshold,
                    minDistX
                );

                if (result) {
                    minDistX = result.dist;
                    snapDeltaX = result.snapDelta;
                    bestXMatch = { 
                        target: result.match.pos, 
                        start: result.match.start,
                        end: result.match.end
                    };
                }
            }
        }
  
        // Horizontal Guides (Vertical movement)
        for (const edge of edgesY) {
            for (const targetVal of tY) {
                const result = checkSnap(edge.val, targetVal, 'horizontal',
                    Math.min(proposedBox.x, target.x),
                    Math.max(proposedBox.x + proposedBox.width, target.x + target.width),
                    threshold,
                    minDistY
                );

                if (result) {
                    minDistY = result.dist;
                    snapDeltaY = result.snapDelta;
                    bestYMatch = { 
                        target: result.match.pos, 
                        start: result.match.start,
                        end: result.match.end
                    };
                }
            }
        }
    }
  
    if (bestXMatch) {
        newGuidelines.push({ type: 'vertical', pos: bestXMatch.target, start: bestXMatch.start, end: bestXMatch.end });
    }
    if (bestYMatch) {
        newGuidelines.push({ type: 'horizontal', pos: bestYMatch.target, start: bestYMatch.start, end: bestYMatch.end });
    }
  
    return {
        translation: { x: snapDeltaX, y: snapDeltaY },
        guidelines: newGuidelines
    };
  };

File: src/snap_logic.rs
use crate::types::{BoundingBox, Guideline, GuidelineType, Point, Polygon};

pub struct SnapResult {
    pub translation: Point,
    pub guidelines: Vec<Guideline>,
}

struct SnapCheck {
    dist: f64,
    snap_delta: f64,
    guideline_type: GuidelineType,
    pos: f64,
    start: f64,
    end: f64,
}

fn check_snap(
    value: f64,
    target: f64,
    guideline_type: GuidelineType,
    start: f64,
    end: f64,
    threshold: f64,
    current_min_dist: f64,
) -> Option<SnapCheck> {
    let dist = (value - target).abs();
    if dist < current_min_dist && dist < threshold {
        Some(SnapCheck {
            dist,
            snap_delta: target - value,
            guideline_type,
            pos: target,
            start,
            end,
        })
    } else {
        None
    }
}

fn parse_polygon_points(points_str: &str) -> Vec<Point> {
    points_str
        .split_whitespace()
        .filter_map(|pair| {
            let coords: Vec<f64> = pair.split(',').filter_map(|s| s.parse().ok()).collect();
            if coords.len() == 2 {
                Some(Point::new(coords[0], coords[1]))
            } else {
                None
            }
        })
        .collect()
}

fn calculate_polygon_bbox(polygon: &Polygon) -> BoundingBox {
    let points = parse_polygon_points(&polygon.points);

    if points.is_empty() {
        return BoundingBox::new(0.0, 0.0, 0.0, 0.0);
    }

    let xs: Vec<f64> = points.iter().map(|p| p.x).collect();
    let ys: Vec<f64> = points.iter().map(|p| p.y).collect();

    let min_x = xs.iter().cloned().fold(f64::INFINITY, f64::min);
    let max_x = xs.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let min_y = ys.iter().cloned().fold(f64::INFINITY, f64::min);
    let max_y = ys.iter().cloned().fold(f64::NEG_INFINITY, f64::max);

    BoundingBox::new(min_x, min_y, max_x - min_x, max_y - min_y)
}

pub fn calculate_snap(
    proposed_box: &BoundingBox,
    polygons: &[Polygon],
    excluded_ids: &[usize],
    canvas_width: f64,
    canvas_height: f64,
    threshold: f64,
) -> SnapResult {
    // Calculate bounding boxes for non-excluded polygons
    let mut other_boxes: Vec<BoundingBox> = polygons
        .iter()
        .enumerate()
        .filter(|(i, _)| !excluded_ids.contains(i))
        .map(|(_, p)| calculate_polygon_bbox(p))
        .collect();

    // Add canvas edges as a bounding box
    other_boxes.push(BoundingBox::new(0.0, 0.0, canvas_width, canvas_height));

    let mut guidelines = Vec::new();
    let mut snap_delta_x = 0.0;
    let mut snap_delta_y = 0.0;
    let mut min_dist_x = threshold;
    let mut min_dist_y = threshold;

    // Edges of the proposed box
    let edges_x = [
        proposed_box.x,                              // start
        proposed_box.x + proposed_box.width / 2.0,   // center
        proposed_box.x + proposed_box.width,         // end
    ];

    let edges_y = [
        proposed_box.y,                              // start
        proposed_box.y + proposed_box.height / 2.0,  // center
        proposed_box.y + proposed_box.height,        // end
    ];

    struct SnapMatch {
        target: f64,
        start: f64,
        end: f64,
    }

    let mut best_x_match: Option<SnapMatch> = None;
    let mut best_y_match: Option<SnapMatch> = None;

    for target_box in &other_boxes {
        let target_x = [
            target_box.x,
            target_box.x + target_box.width / 2.0,
            target_box.x + target_box.width,
        ];

        let target_y = [
            target_box.y,
            target_box.y + target_box.height / 2.0,
            target_box.y + target_box.height,
        ];

        // Vertical guides (horizontal movement)
        for &edge in &edges_x {
            for &target_val in &target_x {
                let start = proposed_box.y.min(target_box.y);
                let end = (proposed_box.y + proposed_box.height).max(target_box.y + target_box.height);

                if let Some(result) = check_snap(
                    edge,
                    target_val,
                    GuidelineType::Vertical,
                    start,
                    end,
                    threshold,
                    min_dist_x,
                ) {
                    min_dist_x = result.dist;
                    snap_delta_x = result.snap_delta;
                    best_x_match = Some(SnapMatch {
                        target: result.pos,
                        start: result.start,
                        end: result.end,
                    });
                }
            }
        }

        // Horizontal guides (vertical movement)
        for &edge in &edges_y {
            for &target_val in &target_y {
                let start = proposed_box.x.min(target_box.x);
                let end = (proposed_box.x + proposed_box.width).max(target_box.x + target_box.width);

                if let Some(result) = check_snap(
                    edge,
                    target_val,
                    GuidelineType::Horizontal,
                    start,
                    end,
                    threshold,
                    min_dist_y,
                ) {
                    min_dist_y = result.dist;
                    snap_delta_y = result.snap_delta;
                    best_y_match = Some(SnapMatch {
                        target: result.pos,
                        start: result.start,
                        end: result.end,
                    });
                }
            }
        }
    }

    if let Some(match_x) = best_x_match {
        guidelines.push(Guideline::new(
            GuidelineType::Vertical,
            match_x.target,
            match_x.start,
            match_x.end,
        ));
    }

    if let Some(match_y) = best_y_match {
        guidelines.push(Guideline::new(
            GuidelineType::Horizontal,
            match_y.target,
            match_y.start,
            match_y.end,
        ));
    }

    SnapResult {
        translation: Point::new(snap_delta_x, snap_delta_y),
        guidelines,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_polygon_points() {
        let points_str = "100,100 200,100 150,200";
        let points = parse_polygon_points(points_str);
        assert_eq!(points.len(), 3);
        assert_eq!(points[0], Point::new(100.0, 100.0));
        assert_eq!(points[1], Point::new(200.0, 100.0));
        assert_eq!(points[2], Point::new(150.0, 200.0));
    }

    #[test]
    fn test_calculate_polygon_bbox() {
        let polygon = Polygon::new(
            "100,100 200,100 150,200".to_string(),
            "#ff0000".to_string(),
            "#000000".to_string(),
            1.0,
        );
        let bbox = calculate_polygon_bbox(&polygon);
        assert_eq!(bbox.x, 100.0);
        assert_eq!(bbox.y, 100.0);
        assert_eq!(bbox.width, 100.0);
        assert_eq!(bbox.height, 100.0);
    }
}
